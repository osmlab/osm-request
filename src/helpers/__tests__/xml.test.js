import fs from 'fs';
import path from 'path';
import {
  buildChangesetXml,
  convertElementXmlToJson,
  convertNotesXmlToJson,
  flattenAttributes,
  xmlToJson,
  jsonToXml,
  encodeXML,
  buildChangesetFromObjectXml,
  buildPreferencesFromObjectXml,
  cleanMapJson,
  convertElementsListXmlToJson,
  convertUserXmlToJson
} from '../xml';

const nodeSample = fs.readFileSync(
  path.join(__dirname, '../../__mocks__/node.xml')
);
const notesSample = fs.readFileSync(
  path.join(__dirname, '../../__mocks__/notes.xml')
);
const wayFullSample = fs.readFileSync(
  path.join(__dirname, '../../__mocks__/way_full.xml')
);
const userSample = fs.readFileSync(
  path.join(__dirname, '../../__mocks__/user.xml')
);

jest.mock('../../../package.json', () => ({
  version: '1.2.3'
}));

describe('XML helpers', () => {
  describe('buildChangesetXml', () => {
    it('Should build a stringified OSM changeset', () => {
      expect(buildChangesetXml('me', 'my comment')).toMatchSnapshot();
      expect(buildChangesetXml()).toMatchSnapshot();
    });

    it('Should handle strings having double quotes', () => {
      expect(
        buildChangesetXml('My "app"', 'Doing some "weird" stuff')
      ).toMatchSnapshot();
    });

    it('Should handle optional tags', () => {
      expect(
        buildChangesetXml('My app', 'Doing some stuff', {
          key_example: 'value example'
        })
      ).toMatchSnapshot();
    });
  });

  describe('buildChangesetFromObjectXml', () => {
    it('Should build a stringified OSM changeset', () => {
      expect(
        buildChangesetFromObjectXml(
          { host: 'server.net', locale: 'fr' },
          'me',
          'my comment'
        )
      ).toMatchSnapshot();
      expect(buildChangesetFromObjectXml()).toMatchSnapshot();
    });

    it('Should handle strings having double quotes', () => {
      expect(
        buildChangesetFromObjectXml(
          { host: 'server.net', locale: 'fr' },
          'My "app"',
          'Doing some "weird" stuff'
        )
      ).toMatchSnapshot();
    });
  });

  describe('buildPreferencesFromObjectXml', () => {
    it('Should build stringified OSM preferences', () => {
      expect(
        buildPreferencesFromObjectXml({ locale: 'fr', color: 'red' })
      ).toMatchSnapshot();
    });
  });

  describe('convertElementXmlToJson', () => {
    it('Should convert an Element XML string into a proper JSON object', async () => {
      const result = await convertElementXmlToJson(
        nodeSample,
        'node',
        '3683625932'
      );
      expect(result).toMatchSnapshot();
    });
  });

  describe('convertElementsListXmlToJson', () => {
    it('Should convert a list of Elements XML string into a proper JSON object', async () => {
      const result = await convertElementsListXmlToJson(wayFullSample, 'node');
      expect(result).toMatchSnapshot();
    });
  });

  describe('convertNotesXmlToJson', () => {
    it('Should convert a Notes XML string into a proper JSON object', async () => {
      const result = await convertNotesXmlToJson(notesSample);
      expect(result).toMatchSnapshot();
    });
  });

  describe('convertUserXmlToJson', () => {
    it('Should convert an User XML string into a proper JSON object', async () => {
      const result = await convertUserXmlToJson(userSample);
      expect(result).toMatchSnapshot();
    });
  });

  describe('flattenAttributes', () => {
    it('Should flatten the weird objects returned by xmlToJson', () => {
      const result = flattenAttributes({
        $: {
          attribute1: 'value1',
          attribute2: 'value2'
        },
        attribute3: ['value3'],
        attribute4: ['value4'],
        undefinedAttribute: undefined,
        emptylist: []
      });

      const expected = {
        attribute1: 'value1',
        attribute2: 'value2',
        attribute3: 'value3',
        attribute4: 'value4'
      };
      expect(result).toEqual(expected);
    });
  });

  describe('xmlToJson', () => {
    it('Should convert an XML string into a JSON object', async () => {
      const result = await xmlToJson(notesSample);
      expect(result).toMatchSnapshot();
    });
  });

  describe('jsonToXml', () => {
    it('Should convert a JSON object into an XML string', () => {
      const result = jsonToXml({
        osm: {
          node: {
            $: {
              attribute1: 'value1',
              attribute2: 'value2'
            },
            attribute3: ['value3'],
            attribute4: ['value4']
          }
        }
      });
      expect(result).toMatchSnapshot();
    });
  });

  describe('encodeXML', () => {
    it('works', () => {
      const result = encodeXML("This is a <weird level='42'>parameter</weird>");
      const expected =
        'This is a &lt;weird level=&apos;42&apos;&gt;parameter&lt;/weird&gt;';
      expect(result).toBe(expected);
    });
  });

  describe('cleanMapJson', () => {
    it('works', () => {
      const mapjson = {
        osm: {
          node: [
            {
              $: { id: '1234', lat: '42.3', lon: '-1.3' },
              tag: [
                {
                  $: {
                    key: 'amenity',
                    val: 'bicycle_parking'
                  }
                }
              ]
            },
            {
              $: { id: '1235', lat: '41.3', lon: '-1.2' }
            }
          ],
          way: [
            {
              $: { id: '456' },
              tag: [{ $: { highway: 'unclassified' } }],
              nd: [{ $: { ref: '1234' } }, { $: { ref: '1235' } }]
            }
          ],
          relation: []
        }
      };
      const result = cleanMapJson(mapjson);
      expect(result).toMatchSnapshot();
    });
  });
});
