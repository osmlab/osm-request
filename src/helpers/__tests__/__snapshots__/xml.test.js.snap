// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XML helpers buildChangesetFromObjectXml Should build a stringified OSM changeset 1`] = `
"
    <osm>
      <changeset>
        <tag k=\\"created_by\\" v=\\"me\\"/>
        <tag k=\\"created_by:library\\" v=\\"OSM Request 2.0.0\\"/>
        <tag k=\\"comment\\" v=\\"my comment\\"/>
        <tag k=\\"host\\" v=\\"server.net\\"/>
        <tag k=\\"locale\\" v=\\"fr\\"/>
      </changeset>
    </osm>
  "
`;

exports[`XML helpers buildChangesetFromObjectXml Should build a stringified OSM changeset 2`] = `
"
    <osm>
      <changeset>
        <tag k=\\"created_by\\" v=\\"\\"/>
        <tag k=\\"created_by:library\\" v=\\"OSM Request 2.0.0\\"/>
        <tag k=\\"comment\\" v=\\"\\"/>
        
      </changeset>
    </osm>
  "
`;

exports[`XML helpers buildChangesetFromObjectXml Should handle strings having double quotes 1`] = `
"
    <osm>
      <changeset>
        <tag k=\\"created_by\\" v=\\"My &quot;app&quot;\\"/>
        <tag k=\\"created_by:library\\" v=\\"OSM Request 2.0.0\\"/>
        <tag k=\\"comment\\" v=\\"Doing some &quot;weird&quot; stuff\\"/>
        <tag k=\\"host\\" v=\\"server.net\\"/>
        <tag k=\\"locale\\" v=\\"fr\\"/>
      </changeset>
    </osm>
  "
`;

exports[`XML helpers buildChangesetXml Should build a stringified OSM changeset 1`] = `
"
    <osm>
      <changeset>
        <tag k=\\"created_by\\" v=\\"me\\"/>
        <tag k=\\"created_by:library\\" v=\\"OSM Request 2.0.0\\"/>
        <tag k=\\"comment\\" v=\\"my comment\\"/>
        
      </changeset>
    </osm>
  "
`;

exports[`XML helpers buildChangesetXml Should build a stringified OSM changeset 2`] = `
"
    <osm>
      <changeset>
        <tag k=\\"created_by\\" v=\\"\\"/>
        <tag k=\\"created_by:library\\" v=\\"OSM Request 2.0.0\\"/>
        <tag k=\\"comment\\" v=\\"\\"/>
        
      </changeset>
    </osm>
  "
`;

exports[`XML helpers buildChangesetXml Should handle strings having double quotes 1`] = `
"
    <osm>
      <changeset>
        <tag k=\\"created_by\\" v=\\"My &quot;app&quot;\\"/>
        <tag k=\\"created_by:library\\" v=\\"OSM Request 2.0.0\\"/>
        <tag k=\\"comment\\" v=\\"Doing some &quot;weird&quot; stuff\\"/>
        
      </changeset>
    </osm>
  "
`;

exports[`XML helpers buildChangesetXml Should handle optional tags 1`] = `
"
    <osm>
      <changeset>
        <tag k=\\"created_by\\" v=\\"My app\\"/>
        <tag k=\\"created_by:library\\" v=\\"OSM Request 2.0.0\\"/>
        <tag k=\\"comment\\" v=\\"Doing some stuff\\"/>
        <tag k=\\"key_example\\" v=\\"value example\\"/>
      </changeset>
    </osm>
  "
`;

exports[`XML helpers buildPreferencesFromObjectXml Should build stringified OSM preferences 1`] = `
"
    <osm>
      <preferences>
        <preference k=\\"locale\\" v=\\"fr\\"/>
        <preference k=\\"color\\" v=\\"red\\"/>
      </preferences>
    </osm>
  "
`;

exports[`XML helpers cleanMapJson works 1`] = `
Object {
  "node": Array [
    Object {
      "$": Object {
        "id": "1234",
        "lat": "42.3",
        "lon": "-1.3",
      },
      "_id": "1234",
      "_type": "node",
      "tag": Array [
        Object {
          "$": Object {
            "key": "amenity",
            "val": "bicycle_parking",
          },
        },
      ],
    },
    Object {
      "$": Object {
        "id": "1235",
        "lat": "41.3",
        "lon": "-1.2",
      },
      "_id": "1235",
      "_type": "node",
    },
  ],
  "relation": Array [],
  "way": Array [
    Object {
      "$": Object {
        "id": "456",
      },
      "_id": "456",
      "_type": "way",
      "nd": Array [
        Object {
          "$": Object {
            "ref": "1234",
          },
        },
        Object {
          "$": Object {
            "ref": "1235",
          },
        },
      ],
      "tag": Array [
        Object {
          "$": Object {
            "highway": "unclassified",
          },
        },
      ],
    },
  ],
}
`;

exports[`XML helpers convertElementXmlToJson Should convert an Element XML string into a proper JSON object 1`] = `
Object {
  "$": Object {
    "changeset": "33150668",
    "id": "3683625932",
    "lat": "44.8331455",
    "lon": "-0.5936602",
    "timestamp": "2015-08-06T09:49:47Z",
    "uid": "2568974",
    "user": "Vinber-Num&Lib",
    "version": "1",
    "visible": "true",
  },
  "_id": "3683625932",
  "_type": "node",
  "tag": Array [
    Object {
      "$": Object {
        "k": "amenity",
        "v": "drinking_water",
      },
    },
  ],
}
`;

exports[`XML helpers convertElementsListXmlToJson Should convert a list of Elements XML string into a proper JSON object 1`] = `
Array [
  Object {
    "$": Object {
      "changeset": "60117678",
      "id": "1531122358",
      "lat": "48.2210040",
      "lon": "-1.7884326",
      "timestamp": "2018-06-24T10:47:29Z",
      "uid": "214436",
      "user": "PanierAvide",
      "version": "2",
      "visible": "true",
    },
    "_id": "1531122358",
    "_type": "node",
  },
  Object {
    "$": Object {
      "changeset": "60117678",
      "id": "1531122360",
      "lat": "48.2210155",
      "lon": "-1.7883932",
      "timestamp": "2018-06-24T10:47:29Z",
      "uid": "214436",
      "user": "PanierAvide",
      "version": "2",
      "visible": "true",
    },
    "_id": "1531122360",
    "_type": "node",
  },
  Object {
    "$": Object {
      "changeset": "60117678",
      "id": "1531122361",
      "lat": "48.2210386",
      "lon": "-1.7883679",
      "timestamp": "2018-06-24T10:47:29Z",
      "uid": "214436",
      "user": "PanierAvide",
      "version": "2",
      "visible": "true",
    },
    "_id": "1531122361",
    "_type": "node",
  },
  Object {
    "$": Object {
      "changeset": "60117678",
      "id": "1531122362",
      "lat": "48.2210245",
      "lon": "-1.7885098",
      "timestamp": "2018-06-24T10:47:29Z",
      "uid": "214436",
      "user": "PanierAvide",
      "version": "2",
      "visible": "true",
    },
    "_id": "1531122362",
    "_type": "node",
  },
  Object {
    "$": Object {
      "changeset": "60117678",
      "id": "1531122366",
      "lat": "48.2210793",
      "lon": "-1.7885211",
      "timestamp": "2018-06-24T10:47:29Z",
      "uid": "214436",
      "user": "PanierAvide",
      "version": "2",
      "visible": "true",
    },
    "_id": "1531122366",
    "_type": "node",
  },
  Object {
    "$": Object {
      "changeset": "60117678",
      "id": "1531122369",
      "lat": "48.2210840",
      "lon": "-1.7883719",
      "timestamp": "2018-06-24T10:47:29Z",
      "uid": "214436",
      "user": "PanierAvide",
      "version": "2",
      "visible": "true",
    },
    "_id": "1531122369",
    "_type": "node",
  },
  Object {
    "$": Object {
      "changeset": "60117678",
      "id": "1531122371",
      "lat": "48.2211072",
      "lon": "-1.7884825",
      "timestamp": "2018-06-24T10:47:29Z",
      "uid": "214436",
      "user": "PanierAvide",
      "version": "2",
      "visible": "true",
    },
    "_id": "1531122371",
    "_type": "node",
  },
  Object {
    "$": Object {
      "changeset": "60117678",
      "id": "1531122372",
      "lat": "48.2211118",
      "lon": "-1.7884258",
      "timestamp": "2018-06-24T10:47:29Z",
      "uid": "214436",
      "user": "PanierAvide",
      "version": "2",
      "visible": "true",
    },
    "_id": "1531122372",
    "_type": "node",
  },
  Object {
    "$": Object {
      "changeset": "60117678",
      "id": "5715015332",
      "lat": "48.2210073",
      "lon": "-1.7884753",
      "timestamp": "2018-06-24T10:47:23Z",
      "uid": "214436",
      "user": "PanierAvide",
      "version": "1",
      "visible": "true",
    },
    "_id": "5715015332",
    "_type": "node",
  },
  Object {
    "$": Object {
      "changeset": "60117678",
      "id": "5715015333",
      "lat": "48.2210508",
      "lon": "-1.7885265",
      "timestamp": "2018-06-24T10:47:23Z",
      "uid": "214436",
      "user": "PanierAvide",
      "version": "1",
      "visible": "true",
    },
    "_id": "5715015333",
    "_type": "node",
  },
  Object {
    "$": Object {
      "changeset": "60117678",
      "id": "5715015334",
      "lat": "48.2210955",
      "lon": "-1.7885055",
      "timestamp": "2018-06-24T10:47:23Z",
      "uid": "214436",
      "user": "PanierAvide",
      "version": "1",
      "visible": "true",
    },
    "_id": "5715015334",
    "_type": "node",
  },
  Object {
    "$": Object {
      "changeset": "60117678",
      "id": "5715015335",
      "lat": "48.2211129",
      "lon": "-1.7884548",
      "timestamp": "2018-06-24T10:47:23Z",
      "uid": "214436",
      "user": "PanierAvide",
      "version": "1",
      "visible": "true",
    },
    "_id": "5715015335",
    "_type": "node",
  },
  Object {
    "$": Object {
      "changeset": "60117678",
      "id": "5715015336",
      "lat": "48.2210616",
      "lon": "-1.7883625",
      "timestamp": "2018-06-24T10:47:23Z",
      "uid": "214436",
      "user": "PanierAvide",
      "version": "1",
      "visible": "true",
    },
    "_id": "5715015336",
    "_type": "node",
  },
  Object {
    "$": Object {
      "changeset": "60117678",
      "id": "5715015337",
      "lat": "48.2211018",
      "lon": "-1.7883943",
      "timestamp": "2018-06-24T10:47:23Z",
      "uid": "214436",
      "user": "PanierAvide",
      "version": "1",
      "visible": "true",
    },
    "_id": "5715015337",
    "_type": "node",
  },
]
`;

exports[`XML helpers convertNotesXmlToJson Should convert a Notes XML string into a proper JSON object 1`] = `
Array [
  Object {
    "close_url": "https://api.openstreetmap.org/api/0.6/notes/1369237/close",
    "comment_url": "https://api.openstreetmap.org/api/0.6/notes/1369237/comment",
    "comments": Array [
      Object {
        "action": "opened",
        "date": "2018-04-22 17:54:49 UTC",
        "html": "<p>Chez Florentin, Faustine, Aimé, Julien et Marine.</p>",
        "text": "Chez Florentin, Faustine, Aimé, Julien et Marine.",
      },
      Object {
        "action": "commented",
        "date": "2018-04-22 17:57:56 UTC",
        "html": "<p>c'est quoi ? une note personnelle ou le nom d'un établissement ?</p>",
        "text": "c'est quoi ? une note personnelle ou le nom d'un établissement ?",
      },
    ],
    "date_created": "2018-04-22 17:54:49 UTC",
    "id": "1369237",
    "lat": "44.4068500",
    "lon": "0.5891000",
    "status": "open",
    "url": "https://api.openstreetmap.org/api/0.6/notes/1369237",
  },
  Object {
    "comments": Array [
      Object {
        "action": "opened",
        "date": "2018-01-16 15:28:34 UTC",
        "html": "<p>carrefour market</p>",
        "text": "carrefour market",
      },
      Object {
        "action": "closed",
        "date": "2018-04-20 11:43:00 UTC",
        "html": "<p></p>",
        "text": "",
        "uid": "2533303",
        "user": "opline",
        "user_url": "https://api.openstreetmap.org/user/opline",
      },
    ],
    "date_closed": "2018-04-20 11:43:00 UTC",
    "date_created": "2018-01-16 15:28:34 UTC",
    "id": "1270165",
    "lat": "44.4890700",
    "lon": "0.1802000",
    "reopen_url": "https://api.openstreetmap.org/api/0.6/notes/1270165/reopen",
    "status": "closed",
    "url": "https://api.openstreetmap.org/api/0.6/notes/1270165",
  },
  Object {
    "close_url": "https://api.openstreetmap.org/api/0.6/notes/1106518/close",
    "comment_url": "https://api.openstreetmap.org/api/0.6/notes/1106518/comment",
    "comments": Array [
      Object {
        "action": "opened",
        "date": "2017-08-16 15:41:56 UTC",
        "html": "<p>Monplaisir</p>",
        "text": "Monplaisir",
      },
      Object {
        "action": "commented",
        "date": "2017-08-16 15:42:43 UTC",
        "html": "<p>Les Carbonnières
</p>",
        "text": "Les Carbonnières
",
      },
      Object {
        "action": "commented",
        "date": "2017-08-17 16:57:49 UTC",
        "html": "<p>lieu-dit, rue ou résidence?</p>",
        "text": "lieu-dit, rue ou résidence?",
        "uid": "379182",
        "user": "Metaltyty",
        "user_url": "https://api.openstreetmap.org/user/Metaltyty",
      },
      Object {
        "action": "commented",
        "date": "2017-08-24 15:44:07 UTC",
        "html": "<p>Lieu dit</p>",
        "text": "Lieu dit",
      },
      Object {
        "action": "commented",
        "date": "2017-10-22 16:06:12 UTC",
        "html": "<p>Monplaisir ou les carbonnières ? Vous nous proposez deux noms</p>",
        "text": "Monplaisir ou les carbonnières ? Vous nous proposez deux noms",
        "uid": "2533303",
        "user": "opline",
        "user_url": "https://api.openstreetmap.org/user/opline",
      },
      Object {
        "action": "commented",
        "date": "2018-04-08 19:35:56 UTC",
        "html": "<p>on cloture, pas de réactions ?</p>",
        "text": "on cloture, pas de réactions ?",
        "uid": "2568974",
        "user": "Vinber-Num&Lib",
        "user_url": "https://api.openstreetmap.org/user/Vinber-Num&Lib",
      },
    ],
    "date_created": "2017-08-16 15:41:56 UTC",
    "id": "1106518",
    "lat": "44.5063613",
    "lon": "0.2216578",
    "status": "open",
    "url": "https://api.openstreetmap.org/api/0.6/notes/1106518",
  },
]
`;

exports[`XML helpers convertUserXmlToJson Should convert an User XML string into a proper JSON object 1`] = `
Object {
  "account_created": "2010-01-02T13:40:42Z",
  "blocks": Object {
    "received": Array [
      Object {
        "active": "0",
        "count": "1",
      },
    ],
  },
  "changesets": Object {
    "count": "3616",
  },
  "contributor-terms": Object {
    "agreed": "true",
  },
  "description": "
  OpenStreetMap passionate, I contribute mainly in the west of France on various subjects : indoor mapping, street equipment, advertisement...
  ",
  "display_name": "PanierAvide",
  "id": "214436",
  "img": Object {
    "href": "https://www.openstreetmap.org/rails/active_storage/representations/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBamNZIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--576881ff3e6c38ff381c9ecceff09761c61951db/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCam9MY21WemFYcGxTU0lOTVRBd2VERXdNRDRHT2daRlZBPT0iLCJleHAiOm51bGwsInB1ciI6InZhcmlhdGlvbiJ9fQ==--efdc66963555b7419037018ba117eb1fea762dc5/Tux_avatar.png",
  },
  "roles": " ",
  "traces": Object {
    "count": "95",
  },
}
`;

exports[`XML helpers jsonToXml Should convert a JSON object into an XML string 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<osm>
  <node attribute1=\\"value1\\" attribute2=\\"value2\\">
    <attribute3>value3</attribute3>
    <attribute4>value4</attribute4>
  </node>
</osm>"
`;

exports[`XML helpers xmlToJson Should convert an XML string into a JSON object 1`] = `
Object {
  "osm": Object {
    "$": Object {
      "generator": "OpenStreetMap server",
      "version": "0.6",
    },
    "note": Array [
      Object {
        "$": Object {
          "lat": "44.4068500",
          "lon": "0.5891000",
        },
        "close_url": Array [
          "https://api.openstreetmap.org/api/0.6/notes/1369237/close",
        ],
        "comment_url": Array [
          "https://api.openstreetmap.org/api/0.6/notes/1369237/comment",
        ],
        "comments": Array [
          Object {
            "comment": Array [
              Object {
                "action": Array [
                  "opened",
                ],
                "date": Array [
                  "2018-04-22 17:54:49 UTC",
                ],
                "html": Array [
                  "<p>Chez Florentin, Faustine, Aimé, Julien et Marine.</p>",
                ],
                "text": Array [
                  "Chez Florentin, Faustine, Aimé, Julien et Marine.",
                ],
              },
              Object {
                "action": Array [
                  "commented",
                ],
                "date": Array [
                  "2018-04-22 17:57:56 UTC",
                ],
                "html": Array [
                  "<p>c'est quoi ? une note personnelle ou le nom d'un établissement ?</p>",
                ],
                "text": Array [
                  "c'est quoi ? une note personnelle ou le nom d'un établissement ?",
                ],
              },
            ],
          },
        ],
        "date_created": Array [
          "2018-04-22 17:54:49 UTC",
        ],
        "id": Array [
          "1369237",
        ],
        "status": Array [
          "open",
        ],
        "url": Array [
          "https://api.openstreetmap.org/api/0.6/notes/1369237",
        ],
      },
      Object {
        "$": Object {
          "lat": "44.4890700",
          "lon": "0.1802000",
        },
        "comments": Array [
          Object {
            "comment": Array [
              Object {
                "action": Array [
                  "opened",
                ],
                "date": Array [
                  "2018-01-16 15:28:34 UTC",
                ],
                "html": Array [
                  "<p>carrefour market</p>",
                ],
                "text": Array [
                  "carrefour market",
                ],
              },
              Object {
                "action": Array [
                  "closed",
                ],
                "date": Array [
                  "2018-04-20 11:43:00 UTC",
                ],
                "html": Array [
                  "<p></p>",
                ],
                "text": Array [
                  "",
                ],
                "uid": Array [
                  "2533303",
                ],
                "user": Array [
                  "opline",
                ],
                "user_url": Array [
                  "https://api.openstreetmap.org/user/opline",
                ],
              },
            ],
          },
        ],
        "date_closed": Array [
          "2018-04-20 11:43:00 UTC",
        ],
        "date_created": Array [
          "2018-01-16 15:28:34 UTC",
        ],
        "id": Array [
          "1270165",
        ],
        "reopen_url": Array [
          "https://api.openstreetmap.org/api/0.6/notes/1270165/reopen",
        ],
        "status": Array [
          "closed",
        ],
        "url": Array [
          "https://api.openstreetmap.org/api/0.6/notes/1270165",
        ],
      },
      Object {
        "$": Object {
          "lat": "44.5063613",
          "lon": "0.2216578",
        },
        "close_url": Array [
          "https://api.openstreetmap.org/api/0.6/notes/1106518/close",
        ],
        "comment_url": Array [
          "https://api.openstreetmap.org/api/0.6/notes/1106518/comment",
        ],
        "comments": Array [
          Object {
            "comment": Array [
              Object {
                "action": Array [
                  "opened",
                ],
                "date": Array [
                  "2017-08-16 15:41:56 UTC",
                ],
                "html": Array [
                  "<p>Monplaisir</p>",
                ],
                "text": Array [
                  "Monplaisir",
                ],
              },
              Object {
                "action": Array [
                  "commented",
                ],
                "date": Array [
                  "2017-08-16 15:42:43 UTC",
                ],
                "html": Array [
                  "<p>Les Carbonnières
</p>",
                ],
                "text": Array [
                  "Les Carbonnières
",
                ],
              },
              Object {
                "action": Array [
                  "commented",
                ],
                "date": Array [
                  "2017-08-17 16:57:49 UTC",
                ],
                "html": Array [
                  "<p>lieu-dit, rue ou résidence?</p>",
                ],
                "text": Array [
                  "lieu-dit, rue ou résidence?",
                ],
                "uid": Array [
                  "379182",
                ],
                "user": Array [
                  "Metaltyty",
                ],
                "user_url": Array [
                  "https://api.openstreetmap.org/user/Metaltyty",
                ],
              },
              Object {
                "action": Array [
                  "commented",
                ],
                "date": Array [
                  "2017-08-24 15:44:07 UTC",
                ],
                "html": Array [
                  "<p>Lieu dit</p>",
                ],
                "text": Array [
                  "Lieu dit",
                ],
              },
              Object {
                "action": Array [
                  "commented",
                ],
                "date": Array [
                  "2017-10-22 16:06:12 UTC",
                ],
                "html": Array [
                  "<p>Monplaisir ou les carbonnières ? Vous nous proposez deux noms</p>",
                ],
                "text": Array [
                  "Monplaisir ou les carbonnières ? Vous nous proposez deux noms",
                ],
                "uid": Array [
                  "2533303",
                ],
                "user": Array [
                  "opline",
                ],
                "user_url": Array [
                  "https://api.openstreetmap.org/user/opline",
                ],
              },
              Object {
                "action": Array [
                  "commented",
                ],
                "date": Array [
                  "2018-04-08 19:35:56 UTC",
                ],
                "html": Array [
                  "<p>on cloture, pas de réactions ?</p>",
                ],
                "text": Array [
                  "on cloture, pas de réactions ?",
                ],
                "uid": Array [
                  "2568974",
                ],
                "user": Array [
                  "Vinber-Num&Lib",
                ],
                "user_url": Array [
                  "https://api.openstreetmap.org/user/Vinber-Num&Lib",
                ],
              },
            ],
          },
        ],
        "date_created": Array [
          "2017-08-16 15:41:56 UTC",
        ],
        "id": Array [
          "1106518",
        ],
        "status": Array [
          "open",
        ],
        "url": Array [
          "https://api.openstreetmap.org/api/0.6/notes/1106518",
        ],
      },
    ],
  },
}
`;
